### PR TITLE
LO-8132: initial centralization of some auth logic into app-tools for shared usage

### PR DESCRIPTION
**Pull request by virdesai**
Initial centralization work for some app-auth to app-tools

### CODE REVIEW

File: src/APIBasedAuth.ts
Comment (virdesai) [inline src/APIBasedAuth.ts:144]: @smolster this sort of stuff is what you had in mind for using https://github.com/lifeomic/one-schema/ , right?

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:8]: Can we remove these ignore statements before merging?

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:25]: Can we make `baseURL` optional, and if it is not defined, assume the production url: `https://apps.us.lifeomic.com/auth/v1/api`

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:29]: What is `ctor`?

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:36]: We seem to be ignoring most of the `options` inputs here. I think we should either:
- not accept the general axios config in the constructor, or
- spread the config into this object

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:53]: The APIs that this package is going to hit will be all unauthenticated endpoints, I think. Do we need this interceptor?

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:61]: Will `this.session` always have the up-to-date value from storage here?

File: test/APIBasedAuth.test.ts
Comment (jkdowdle) [inline test/APIBasedAuth.test.ts:7]: Huge nit, I typically like to explicitly declare each variable with separate `let` keyword

**Review by smolster — COMMENTED** [review]
This looks like a great start! @mdlavin might be interested in reviewing too.

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:29]: no clue, it was in there for the LOAuth so didn't rename it

File: src/APIBasedAuth.ts
Comment (jkdowdle) [inline src/APIBasedAuth.ts:61]: In the case that `session` needs to come from some other storage (user clicks login button and opens new tab for example) would the caller need to handle that explicitly with `baseInput`?

Comment (smolster) [reply src/APIBasedAuth.ts:61]: The motivation behind storing the session is primarily this flow:
1. Someone triggers passwordless.
2. They close their web page / kill their app.
3. They click on the link in their passwordless email, which re-opens their app.
4. Now, we need the `session` from step 1, which is no longer in memory (it's only in storage).

So, it's possible that `confirmPasswordlessAuth` might get called before `initiatePasswordlessAuth` is ever called (in a particular session). That makes me think we need to fetch from storage here.

**Review by jkdowdle — APPROVED** [review]

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:8]: I can, just need to figure out more tests then, it's just ignoring for coverage thresholds right now

File: src/APIBasedAuth.ts
Comment (smolster) [reply src/APIBasedAuth.ts:36]: If we go for Option 1, we get the benefit of not exposing any `axios`-specific API, which would be cool.

File: src/APIBasedAuth.ts
Comment (smolster) [reply src/APIBasedAuth.ts:29]: +1 for removing it.

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts]: within TS I've made the config pick only baseURL and withCredentials from axios

**Review by jkdowdle — DISMISSED** [review]

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:9]: this one may not have coverage issues anymore cause I'm a good amount over the threshold but using it in the tests files is the only reason I'm exporting it currently from this file

**Comment by loscm** [comment]
Requested review from a random `@lifeomic/security` team member, @bishopb, due to term: **auth**.

This review is not blocking and is for broader awareness.  Consider if this change requires deeper security review and ask when it is necessary.

- [Automation](https://github.com/lifeomic/pr-security-flag-scanner/blob/master/src/lambdas/pr-security-flag-scanner/index.ts) | [#pr-flagged-sec-review](https://lifeomic.slack.com/messages/pr-flagged-sec-review)

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:121]: Two questions:
1. Why is `withCredentials` included here?
2. Can we either:
  - avoid picking from the `AxiosRequestConfig` type, or
  - include `@types/axios` in our `dependencies` array?

If we export references to AxiosRequestConfig from our type declarations, type compilation will fail when someone tries to use this without `@types/axios` in their dependencies. It should go in ours.

Comment (virdesai) [reply src/APIBasedAuth.ts:121]: 1. it's there cause I see we're using it for preventia in ClientUI
2. yeah, I'll just type it individually instead of adding the lib

Comment (smolster) [reply src/APIBasedAuth.ts]: > it's there cause I see we're using it for preventia in ClientUI

You're right that the flag is specified, but it's statically set to `false` for the client that interacts with the auth api. `false` is the default setting -- we should be able to leave it off here.

@jkdowdle can confirm

File: src/APIBasedAuth.ts
Comment (aecorredor) [reply src/APIBasedAuth.ts:7]: It was mostly a nit. The main benefit I see is not having to leave those explicit comments, as having to export them just for testing purposes signals that they method is meant to be used as a helper that's part of another module, if not, the testing code wouldn't really care about it, right?

**Review by aecorredor — APPROVED** [review]

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts]: I'll change that in my follow up with storage

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:7]: if you ignored helper functions from testing code coverage then yeah

**Review by jkdowdle — APPROVED** [review]

File: src/APIBasedAuth.ts
Comment (mdlavin) [reply src/APIBasedAuth.ts:40]: @virdesai did you have a thought about this? It seems like the flow that Swain pointed it is right, will a session exist at this point?

File: src/APIBasedAuth.ts
Comment (mdlavin) [inline src/APIBasedAuth.ts:42]: I see here an other places a lot of `.then` and `.catch` are used. Is that intentional rather than use async/await ?

File: src/APIBasedAuth.ts
Comment (mdlavin) [reply src/APIBasedAuth.ts:40]: Opps. I see that instead of fixing it here you opened a new PR. I'll review there

**Pull request merged**
